File: rag-test-py/src/loader_and_splitter2.py
"""loader_and_splitter2.py - 网页加载、切分并执行 RAG 检索"""
import asyncio
import os

from dotenv import load_dotenv
from langchain_community.document_loaders import WebBaseLoader
from langchain_core.vectorstores import InMemoryVectorStore
from langchain_openai import ChatOpenAI, OpenAIEmbeddings
from langchain_text_splitters import RecursiveCharacterTextSplitter


load_dotenv()


async def main() -> None:
    model = ChatOpenAI(
        temperature=0,
        model=os.getenv("MODEL_NAME", "qwen-coder-turbo"),
        api_key=os.getenv("OPENAI_API_KEY"),
        base_url=os.getenv("OPENAI_BASE_URL"),
    )
    embeddings = OpenAIEmbeddings(
        api_key=os.getenv("OPENAI_API_KEY"),
        model=os.getenv("EMBEDDINGS_MODEL_NAME", "text-embedding-v3"),
        base_url=os.getenv("OPENAI_BASE_URL"),
        check_embedding_ctx_length=False,
    )

    # 每个 URL 对应一个 Document
    loader = WebBaseLoader(web_path="https://juejin.cn/post/7233327509919547452")
    documents = loader.load()
    assert len(documents) >= 1

    print(f'documents长度{len(documents)}')

    # 让流程更稳定、可控，它先只取第一篇文档
    source_document = documents[0]
    source_document.page_content = "\n".join(
        line.strip() for line in source_document.page_content.splitlines() if line.strip()
    )
    print(f"Total characters: {len(source_document.page_content)}")

    '''
    overloap 只有文本超过 chunk size，文本被打断了才会加，不是所有的块都会有 overlap
    '''
    text_splitter = RecursiveCharacterTextSplitter(
        chunk_size=500,
        chunk_overlap=50, # 通常设置为 chunkSize 的 10% - 20%
        separators=["。", "！", "？"],
    )
    split_documents = text_splitter.split_documents([source_document])

    print(split_documents)
    print(f"文档分割完成，共 {len(split_documents)} 个分块\n")

    print("正在创建向量存储...")
    vector_store = InMemoryVectorStore(embeddings)
    await vector_store.aadd_documents(split_documents)
    print("向量存储创建完成\n")

    question = "父亲的去世对作者的人生态度产生了怎样的根本性逆转？"
    print("=" * 80)
    print(f"问题: {question}")
    print("=" * 80)

    scored_results = await vector_store.asimilarity_search_with_score(question, k=2)
    retrieved_docs = [doc for doc, _ in scored_results]

    print("\n【检索到的文档及相似度评分】")
    for index, (doc, score) in enumerate(scored_results, start=1):
        print(f"\n[文档 {index}] 相似度: {1 - score:.4f}")
        print(f"内容: {doc.page_content}")
        if doc.metadata:
            print(f"元数据: {doc.metadata}")

    context = "\n\n━━━━━\n\n".join(
        f"[片段{index}]\n{doc.page_content}" for index, doc in enumerate(retrieved_docs, start=1)
    )
    prompt = (
        "你是一个文章辅助阅读助手，根据文章内容来解答：\n\n"
        f"文章内容：\n{context}\n\n问题: {question}\n\n你的回答:"
    )

    print("\n【AI 回答】")
    response = await model.ainvoke(prompt)
    print(response.content)
    print()


if __name__ == "__main__":
    asyncio.run(main())

File: rag-test/src/loader-and-splitter2.mjs
import "dotenv/config";
import "cheerio";
import { ChatOpenAI, OpenAIEmbeddings } from "@langchain/openai";
import { RecursiveCharacterTextSplitter } from "@langchain/textsplitters";
import { MemoryVectorStore } from "@langchain/classic/vectorstores/memory";
import { CheerioWebBaseLoader } from "@langchain/community/document_loaders/web/cheerio";

const model = new ChatOpenAI({
  temperature: 0,
  model: process.env.MODEL_NAME,
  apiKey: process.env.OPENAI_API_KEY,
  configuration: {
    baseURL: process.env.OPENAI_BASE_URL,
  },
});

const embeddings = new OpenAIEmbeddings({
  apiKey: process.env.OPENAI_API_KEY,
  model: process.env.EMBEDDINGS_MODEL_NAME,
  configuration: {
    baseURL: process.env.OPENAI_BASE_URL
  },
});

const cheerioLoader = new CheerioWebBaseLoader(
  "https://juejin.cn/post/7233327509919547452",
  {
    selector: '.main-area p'
  }
);

const documents = await cheerioLoader.load();

console.assert(documents.length === 1);
console.log(`Total characters: ${documents[0].pageContent.length}`);

const textSplitter = new RecursiveCharacterTextSplitter({
  chunkSize: 500,  // 每个分块的字符数
  chunkOverlap: 50,  // 分块之间的重叠字符数
  separators: ["。", "！", "？"],  // 分割符，优先使用段落分隔
});

const splitDocuments = await textSplitter.splitDocuments(documents);

console.log(splitDocuments);


console.log(`文档分割完成，共 ${splitDocuments.length} 个分块\n`);

console.log("正在创建向量存储...");
const vectorStore = await MemoryVectorStore.fromDocuments(
  splitDocuments,
  embeddings,
);
console.log("向量存储创建完成\n");

const retriever = vectorStore.asRetriever({ k: 2 });

const questions = [
  "父亲的去世对作者的人生态度产生了怎样的根本性逆转？"
];

// RAG 流程：对每个问题进行检索和回答
for (const question of questions) {
  console.log("=".repeat(80));
  console.log(`问题: ${question}`);
  console.log("=".repeat(80));
  
  // 使用 similaritySearchWithScore 获取文档和相似度评分（一次调用即可）
  const scoredResults = await vectorStore.similaritySearchWithScore(question, 2);
  
  // 从 scoredResults 中提取文档和评分
  const retrievedDocs = scoredResults.map(([doc]) => doc);
  
  // 打印检索到的文档和相似度评分
  console.log("\n【检索到的文档及相似度评分】");
  scoredResults.forEach(([doc, score], i) => {
    const similarity = (1 - score).toFixed(4);
    
    console.log(`\n[文档 ${i + 1}] 相似度: ${similarity}`);
    console.log(`内容: ${doc.pageContent}`);
    if (doc.metadata && Object.keys(doc.metadata).length > 0) {
      console.log(`元数据:`, doc.metadata);
    }
  });
  
  // 构建 prompt
  const context = retrievedDocs
    .map((doc, i) => `[片段${i + 1}]\n${doc.pageContent}`)
    .join("\n\n━━━━━\n\n");
  
  const prompt = `你是一个文章辅助阅读助手，根据文章内容来解答：

文章内容：
${context}

问题: ${question}

你的回答:`;
  
  console.log("\n【AI 回答】");
  const response = await model.invoke(prompt);
  console.log(response.content);
  console.log("\n");
}
